File: sysprog-uebungen/aufgabe5/notenspiegel.cpp
/*
 * notenspiegel.cpp
 *
 * Liest die Namen von Faechern mit den zugehoerigen Noten in eine
 * verkettete Liste ein und gibt dann einen Notenspiegel aus.
 *
 * Autor: H.Drachenfels
 * Version: 19.12.2019
 */
#include "benotung.h"
#include "fachnote.h"
#include "fachnoten_liste.h"
#include <iostream>
#include <iomanip>
#include <stdexcept>
#include <cstdio>
#include <string>

namespace
{
    std::ostream& operator<<(std::ostream& os, const benotung& n);
    void delete_fachnote(fachnote *);
}

int main(void)
{
    fachnoten_liste liste(delete_fachnote); // leere Liste

    //--------------------------------------------- Notenspiegel einlesen
    std::cerr << "Fach und Note zwischen "
              << benotung::beste.int_value()
              << " und "
              << benotung::schlechteste.int_value()
              << " eingeben (Ende mit Strg-D):\n";

    while (true)
    {
        try
        {
            // Fach und Note einlesen:
            std::string fach;
            if (!(std::cin >> fach))
            {
                break;
            }

            int note;
            if (!(std::cin >> note))
            {
                if (std::cin.eof())
                {
                    std::cerr << "Fach " << fach << " ohne Note ignoriert!\n";
                    break;
                }

                std::cin.clear();
                std::string s;
                std::cin >> s;
                std::cerr << "Eingabefehler: " << s << '\n';
                continue;
            }

            // neue Fachnote in Notenliste eintragen:
            fachnote *f = new fachnote{fach, benotung{note}};
            liste.insert(f);
        }
        catch (std::invalid_argument& x)
        {
            std::cerr << "Eingabefehler: " << x.what() << '\n';
            continue;
        }
    }

    //--------------------------------------------- Notenspiegel ausgeben
    std::size_t max_laenge = 0;
    for (fachnote *f : liste)
    {
        if (max_laenge < f->fach.length())
        {
            max_laenge = f->fach.length();
        }
    }

    std::cout << "\nNOTENSPIEGEL\n";
    for (fachnote *f : liste)
    {
        std::cout << std::setw(max_laenge) << std::left << f->fach << '\t'
                  << f->note << '\t';
        if (f->note == benotung::beste)
        {
            std::cout << "mit Bestnote bestanden\n";
        }
        else if (f->note.ist_bestanden())
        {
            std::cout << "bestanden\n";
        }
        else
        {
            std::cout << "nicht bestanden\n";
        }
    }
}

namespace
{
    std::ostream& operator<<(std::ostream& os, const benotung& n)
    {
        int note = n.int_value();
        os << note / 10 << ',' << note % 10;
        return os;
    }

    void delete_fachnote(fachnote *f)
    {
        delete f;
    }
}

File: Semester 1/prog1-uebungen/aufgabe5/Notenspiegel.java
// Notenspiegel.java
package aufgabe5;
import java.util.NoSuchElementException;
import java.util.Scanner;

/**
 * Notenspiegel liest die Namen von F&auml;chern mit den zugeh&ouml;rigen Noten
 * in eine verkettete Liste ein und gibt dann einen Notenspiegel aus.
 * @author Dominik Nowara
 * @version 1.0
 */
public final class Notenspiegel {
    private Notenspiegel() { }

    private static final Scanner EINGABE = new Scanner(System.in);

    /**
     * main ist der Startpunkt des Programms.
     * @param args wird nicht verwendet.
     */
    public static void main(String[] args) {
        FachnotenListe liste = new FachnotenListe(); // leere Liste

        //--------------------------------------------- Notenspiegel einlesen
        System.err.printf(
            "Faecher mit Noten zwischen %d und %d eingeben "
            + "(Ende mit Strg-D):%n",
            Note.BESTE.intValue(), Note.SCHLECHTESTE.intValue());

        while (EINGABE.hasNext()) {
            try {
                //------------------------------------ Fach und Note einlesen

                String notenWert = EINGABE.next();

                if (EINGABE.hasNextInt()) {
                    var note = Note.valueOf(EINGABE.nextInt());
                    liste.insert(new Fachnote(notenWert, note));
                } else {
                    var note = Note.valueOf(EINGABE.next());
                    liste.insert(new Fachnote(notenWert, note));
                }

                //--------------------- neue Fachnote in Notenliste eintragen


            } catch (IllegalArgumentException x) {
                System.err.printf("Eingabefehler: %s%n", x.getMessage());
                continue;
            } catch (NoSuchElementException x) {
                System.err.println("Fach ohne Note ignoriert!");
                break;
            }
        }

        //--------------------------------------------- Notenspiegel ausgeben

        final int zehn = 10;

        System.out.println("NOTENSPIEGEL");
        var i = liste.new Iterator();
        while (i.hasNext()) {
            var next = i.next();

            var notenAusgabe = (next.note.istBestanden())
                    ? "bestanden" : "nicht bestanden";
            if (next.note.intValue() == zehn) {
                notenAusgabe = "mit Bestnote bestanden";
            }

            System.out.println(String.format("%-25s", next.fach)
                + String.format("%-5s", next.note.toString())
                + notenAusgabe);
        }

    } // main
}

